### PR TITLE
[GraphBolt][CUDA] Pipelined sampling optimization

### PR DESCRIPTION
**Pull request by mfbalin**
## Description

We overlap the UVA graph accesses with the rest of the computations by launching them in the UVA stream. We use a separate thread becase the `index_select_csc` is blocking, unlike `index_select`, which is nonblocking. I suspect the GIL in python is making multithreading inefficient a little bit, so more performance can be unlocked in the future after python with no GIL is available, see https://peps.python.org/pep-0703/#gpu-heavy-workloads-require-multi-core-processing

I am getting %10-%20 end-to-end speedup for `sample_layer_neighbor`. `sample_neighbor` is currently not getting much speedup because it does not need to access the whole indices tensor, which we are fetching through this optimization. Will need to specialize for `sample_neighbor` later.

This optimization will be the core of us providing the users a way to define their own custom samplers using `dgl.sparse`. My next PR will show how to implement LADIES while still taking advantage of the GraphBolt optimizations.

image

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/src/cuda/index_select_csc_impl.cu
Comment (frozenbugs) [inline graphbolt/src/cuda/index_select_csc_impl.cu:1]: split out the cu change to separate PR.

File: python/dgl/graphbolt/dataloader.py [hunk 1]
```diff
@@ -226,6 +258,30 @@ def __init__(
...
+            and torch.cuda.is_available()
+        ):
+            torch.ops.graphbolt.set_max_uva_threads(max_uva_threads)
+            samplers = dp_utils.find_dps(
+                datapipe_graph,
+                SamplePerLayer,
+            )
+            executor = ThreadPoolExecutor(max_workers=1)
+            for sampler in samplers:
+                datapipe_graph = dp_utils.replace_dp(
```
Comment (frozenbugs) [inline python/dgl/graphbolt/dataloader.py:273]: It is a little bit weird to see FetcherAndSampler in the dataloader.py, but not sure what's the best way to do. I am worried about we can't maintain the complexity with all this dp replacement.

Is it possible to code this in neighbor_sampler.py in some way?

Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py:273]: It will lead to circular dependency in that case. I can remove FetcherAndSampler if you want and manually chain the datapipes here.

Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py:270]: Pushed my changes.

File: graphbolt/src/cuda/index_select_csc_impl.cu
Comment (mfbalin) [reply graphbolt/src/cuda/index_select_csc_impl.cu:1]: #7056

File: python/dgl/graphbolt/dataloader.py (hunk 1 above)
Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py:270]: Or, we can move the Waiter and Bufferer into a different file so that dataloader and the neighborsampler can import it.

Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py:190]: Moved the datapipe components into base.py and the FetcherAndSampler into sample_neighbors.py.

**Comment by mfbalin** [comment]
@frozenbugs I am hoping that with my most recent commits, the main concerns related to your previous reviews should be addressed now.

File: python/dgl/graphbolt/impl/neighbor_sampler.py [hunk 2]
```diff
@@ -1,18 +1,125 @@
...
+        super().__init__(datapipe, self._fetch_per_layer)
+        self.graph = sample_per_layer_obj.sampler.__self__
+        self.prob_name = sample_per_layer_obj.prob_name
+        self.stream = stream
+        if executor is None:
+            self.executor = ThreadPoolExecutor(max_workers=1)
+        else:
+            self.executor = executor
+
+    def _fetch_per_layer_helper(self, minibatch, stream):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/neighbor_sampler.py:37]:
```suggestion
    def _fetch_per_layer_impl(self, minibatch, stream):
```

File: python/dgl/graphbolt/impl/neighbor_sampler.py [hunk 3]
```diff
@@ -1,18 +1,125 @@
...
 from .sampled_subgraph_impl import SampledSubgraphImpl
 
 
-__all__ = ["NeighborSampler", "LayerNeighborSampler"]
+__all__ = ["NeighborSampler", "LayerNeighborSampler", "SamplePerLayer"]
+
+
+@functional_datapipe("fetch_insubgraph_data")
+class FetchInsubgraphData(Mapper):
+    """Fetches the insubgraph and wraps it in a FusedCSCSamplingGraph object."""
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/neighbor_sampler.py:23]: Worth to clarify what does this function do. it fetch insubgraph neighbors for seed nodes, but for probs_or_mask are you only get for seed nodes or all insubgraph neighbors?

File: python/dgl/graphbolt/impl/neighbor_sampler.py [hunk 4]
```diff
@@ -1,18 +1,125 @@
...
+                if stream is not None and tensor.is_cuda:
+                    tensor.record_stream(stream)
+
+            indptr, indices = index_select(self.graph.indices, index, None)
+            record_stream(indptr)
+            record_stream(indices)
+            output_size = len(indices)
+            if self.graph.type_per_edge is not None:
+                _, type_per_edge = index_select(
+                    self.graph.type_per_edge, index, output_size
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/neighbor_sampler.py:56]: Is this correct to use index or indices?

File: python/dgl/graphbolt/impl/neighbor_sampler.py (hunk 3 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/impl/neighbor_sampler.py:23]: Also please add unit test for this method.

**Comment by frozenbugs** [comment]
Overall LGTM, questions on python/dgl/graphbolt/impl/neighbor_sampler.py.

File: python/dgl/graphbolt/impl/neighbor_sampler.py (hunk 4 above)
Comment (mfbalin) [reply python/dgl/graphbolt/impl/neighbor_sampler.py:56]: Index because we defined `index_select` above as:
```python
            index_select = partial(
                torch.ops.graphbolt.index_select_csc, self.graph.csc_indptr
            )
```

Comment (mfbalin) [reply python/dgl/graphbolt/impl/neighbor_sampler.py:73]: We are getting the vaue for the insubgraph neighbors of seed nodes when used like this. Do you think I can make this more clear in the code?

Comment (frozenbugs) [reply python/dgl/graphbolt/impl/neighbor_sampler.py:73]: I see, index_select is a reserved name for the specific op, reuse it here causes confusion. Consider rename it to index_select_csc_with_indptr or better ones.

**Review by frozenbugs — APPROVED** [review]

**Comment by mfbalin** [comment]
@dgl-bot

**Comment by mfbalin** [comment]
@dgl-bot

**Comment by mfbalin** [comment]
@Rhett-Ying Does the CI have a problem?

**Comment by mfbalin** [comment]
@dgl-bot

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Comment by Rhett-Ying** [comment]
@dgl-bot
